File: Algorithm-go/reverseLinkedList.go
package main

import "fmt"

// 节点
type ListNode struct {
  Val int
  Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
		var pre, cur *ListNode
		cur = head
		for cur != nil {
			tem := cur.Next
			cur.Next = pre
			pre, cur = cur, tem
		}
		return pre
}

func runReverseList() {
	var head = ListNode{Val: 0}
	var list = &head
	index := 0
	for i := 0; i < 10; i ++ {
		fmt.Println("index: ", index,"value: ", head.Val)
		l := ListNode{Val: i}
		list.Next = &l
		list = &l
		index ++
	}

	r := reverseList(&head)
	fmt.Println("\n Reversed: ")
	index2 := 0
	for r != nil {
		fmt.Println("value: ", r.Val, "index: ", index2)
		index2 ++
		r = r.Next
	}
}

File: Algorithm/Algorithm/ReverseLinkedList.swift
//
//  reverse linked list.swift
//  Algorithm
//
//  Created by Ye Yongping on 2019/8/26.
//  Copyright © 2019 Yeyongping. All rights reserved.
//

import Foundation

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
}

func reverseList(_ head: ListNode?) -> ListNode? {
    var (pre, cur): (ListNode?, ListNode?) = (nil, head)
    while cur != nil {
        let next = cur!.next
        cur!.next = pre
        pre = cur
        cur = next
    }
    return pre
}

func runReverseLinkedList () {
    let head = ListNode(0)
    var list = head
    for i in 1..<10 {
        let l = ListNode(i)
        list.next = l
        list = l
    }
    
    var h: ListNode? = head
    var j = 0
    repeat {
        print("value: ",h!.val, "  index: ", j)
        j += 1
        h = h?.next
    } while h?.next != nil
    print("\n")
    
    h = reverseList(head)
    
    var index = 0
    repeat {
        print("value: ",h!.val, "  index: ", index)
        index += 1
        h = h?.next
    } while h?.next != nil
}

